pkg/relayer/proxy/relay_meter.go

```diff
@@ -0,0 +1,384 @@
+		appMetrics = &appRelayMeter{
+			app:                    app,
+			consumedCoin:           cosmostypes.NewInt64Coin(volatile.DenomuPOKT, 0),
+			maxCoin:                maxAmount,
+			sessionHeader:          reqMeta.SessionHeader,
+			sharedParams:           sharedParams,
+			service:                &service,
+			serviceRelayDifficulty: serviceRelayDifficulty,
+		}
+
+		rmtr.supplierToAppMetricsMap[supplierAddress][appAddress] = appMetrics
+	}
+
+	return appMetrics, nil
+}
+
+// filterTypedEvents filters the provided events bytes for the typed event T.
+// T is then filtered by the provided filter function.
+func filterTypedEvents[T proto.Message](
+	ctx context.Context,
+	eventBzObs client.EventsBytesObservable,
+	filterFn func(T) bool,
+) observable.Observable[T] {
+	eventObs, eventCh := channel.NewObservable[T]()
+	channel.ForEach(ctx, eventBzObs, func(ctx context.Context, maybeTxBz either.Bytes) {
+		if maybeTxBz.IsError() {
+			return
+		}
+		txBz, _ := maybeTxBz.ValueOrError()
+
+		// Try to deserialize the provided bytes into an abci.TxResult.
+		txResult, err := tx.UnmarshalTxResult(txBz)
+		if err != nil {
+			return
+		}
+
+		for _, event := range txResult.Result.Events {
+			eventApplicationStakedType := cosmostypes.MsgTypeURL(*new(T))
+			if strings.Trim(event.GetType(), "/") != strings.Trim(eventApplicationStakedType, "/") {
+				continue
+			}
+
+			typedEvent, err := cosmostypes.ParseTypedEvent(event)
+			if err != nil {
+				return
+			}
+
+			castedEvent, ok := typedEvent.(T)
+			if !ok {
+				return
+			}
+
+			// Apply the filter function to the typed event.
+			if filterFn == nil || filterFn(castedEvent) {
+				eventCh <- castedEvent
+				return
+			}
+		}
+	})
+
+	return eventObs
+}
+
+// getMinedRelayCostCoin returns the cost of a relay based on the shared parameters and the service.
+// relayCost = CUPR * CUTTM * relayDifficultyMultiplier
+func getMinedRelayCostCoin(
+	sharedParams *sharedtypes.Params,
+	service *sharedtypes.Service,
+	relayMiningDifficulty servicetypes.RelayMiningDifficulty,
+) (cosmostypes.Coin, error) {
+	// Get the difficulty multiplier based on the relay mining difficulty.
+	difficultyTargetHash := relayMiningDifficulty.GetTargetHash()
+	difficultyMultiplier := protocol.GetRelayDifficultyMultiplier(difficultyTargetHash)
+
+	// Get the estimated cost of the relay if it gets mined.
+	relayCostAmt := service.ComputeUnitsPerRelay * sharedParams.GetComputeUnitsToTokensMultiplier()
+	relayCostRat := big.NewRat(int64(relayCostAmt), 1)
+	estimatedRelayCostRat := big.NewRat(0, 1).Mul(relayCostRat, difficultyMultiplier)
+	estimatedRelayCost := big.NewInt(0).Quo(estimatedRelayCostRat.Num(), estimatedRelayCostRat.Denom())
+
+	return cosmostypes.NewCoin(volatile.DenomuPOKT, math.NewIntFromBigInt(estimatedRelayCost)), nil
+}
+
+// getSupplierAppStake returns the portion of the application stake that can be consumed
+// per supplier per session.
+func getSupplierAppStake(
```

Comment:
Wdyt of `getAppStakePortionPayableToSupplier`?

I know it's longer, but becomes much clearer. I think it's what led me to be confused about some of the other comments I left.